File: code/autoencoder.tfidf.py
import os, pandas
h2o.init()

#If next line fails, instead set path to location of datasets.
path = os.path.dirname(__file__)
tfidf = h2o.import_file( os.path.join(
  path, "../datasets/movie.tfidf.csv") )

# This is the minimal model: 563 --> 2 --> 563. It trains in under a second, and gives MSE of 0.037
m = h2o.estimators.deeplearning.H2OAutoEncoderEstimator(
  hidden=[2], seed = 707
  )
m.train(x=range(1,564), training_frame=tfidf)
f = m.deepfeatures(tfidf,layer=0)


# Here is the stacked auto-encoder
m1 = h2o.estimators.deeplearning.H2OAutoEncoderEstimator(
  hidden = [128,64,11,64,128], seed = 707, activation = "Tanh", epochs = 400
  )
m1.train(x = range(1,564), training_frame = tfidf)
f1 = m1.deepfeatures(tfidf, layer = 2)

m2 = h2o.estimators.deeplearning.H2OAutoEncoderEstimator(
  hidden = [2], seed = 707, activation = "Tanh", epochs = 400
  )
m2.train(x = range(0,11), training_frame = f1)
f2 = m2.deepfeatures(f1, layer = 0)

File: code/autoencoder.tfidf.R
library(h2o)
h2o.init(nthreads = -1)

tfidf <- h2o.importFile("../datasets/movie.tfidf.csv")

m <- h2o.deeplearning(2:564, training_frame = tfidf,
  hidden = c(2), autoencoder = T, activation = "Tanh")
f <- h2o.deepfeatures(m, tfidf, layer = 1)

#Print it
d <- as.matrix(f[1:30,])
labels <- as.vector(tfidf[1:30,1])
plot(d, pch = 17)
text(d, labels, pos = 3)


#Plot for book (1600x1100)
plot(d, cex=1.0, pch=17, xlim=range(d[,1])*1.2, ylim=range(d[,2])*1.05, xlab="",ylab="",cex.lab=1.75, cex.axis=1.75)
text(d,labels = labels,cex=1.75,pos=3, offset = 0.33)

#An alternative way to alter xlim and ylim
plot(d, cex=1.0, pch=17, xlim=range(d[,1])+c(-0.1,0.1), ylim=range(d[,2])+c(-0.02,0.02), xlab="",ylab="",cex.lab=1.75, cex.axis=1.75)
text(d,labels = labels,cex=1.75,pos=3, offset = 0.33)

